File: test-ts-full/src/tools/databaseTool.ts
/**
 * Database Query Tool (Template)
 *
 * Execute database queries with parameterization and validation.
 * NOTE: Install your database driver (pg, mysql2, etc.) and configure connection.
 */

import { z } from 'zod';
import { recordToolCall } from '../observability/metrics.js';
import { logger } from '../observability/logger.js';

// TODO: Install and import your database driver
// import { Pool } from 'pg';

const QuerySchema = z.object({
  query: z.string().describe('SQL query (use parameterized queries)'),
  params: z.array(z.any()).optional().describe('Query parameters'),
  timeout: z.number().default(30000).describe('Query timeout in ms')
});

export type QueryInput = z.infer<typeof QuerySchema>;

export interface QueryOutput {
  rows: any[];
  rowCount: number;
  duration: number;
}

// TODO: Configure your database connection
// const pool = new Pool({
//   connectionString: process.env.DATABASE_URL
// });

export async function databaseQueryTool(input: QueryInput): Promise<QueryOutput> {
  const { query, params = [], timeout } = input;

  const startTime = Date.now();
  logger.info(`Executing database query`);

  // Validate query type (prevent destructive operations if needed)
  const forbiddenKeywords = ['DROP', 'TRUNCATE', 'DELETE FROM'];
  const upperQuery = query.toUpperCase();

  for (const keyword of forbiddenKeywords) {
    if (upperQuery.includes(keyword)) {
      throw new Error(`Query contains forbidden keyword: ${keyword}`);
    }
  }

  try {
    // TODO: Implement actual database query
    // const client = await pool.connect();
    // const result = await client.query({ text: query, values: params, query_timeout: timeout });
    // client.release();

    // Placeholder implementation
    const result = {
      rows: [],
      rowCount: 0
    };

    const duration = Date.now() - startTime;
  recordToolCall('database_query', duration, true);
  logger.info(`Query completed in ${duration}ms`);

    return {
      rows: result.rows,
      rowCount: result.rowCount,
      duration: 0
    };
  } catch (error) {
    throw new Error(`Database query failed: ${error.message}`);
  }
}

export const databaseToolSchema = {
  name: 'database_query',
  description: 'Execute database queries with validation (configure connection first)',
  parameters: QuerySchema
};

File: test-py-full/src/tools/database_tool.py
"""
Database Query Tool (Template)

Execute database queries with parameterization and validation.
NOTE: Install your database driver (psycopg2, pymysql, etc.) and configure connection.
"""

from typing import Dict, Any, List, Optional, Any
from datetime import datetime
from pydantic import BaseModel, Field
from src.observability.metrics import record_tool_call
from src.observability.logger import logger

# TODO: Install and import your database driver
# import asyncpg

class QueryInput(BaseModel):
    query: str = Field(..., description="SQL query (use parameterized queries)")
    params: Optional[List[Any]] = Field(default=None, description="Query parameters")
    timeout: int = Field(default=30, description="Query timeout in seconds")

class QueryOutput(BaseModel):
    rows: List[Dict[str, Any]]
    row_count: int
    duration: float

# TODO: Configure your database connection
# DATABASE_URL = os.getenv("DATABASE_URL")
# pool = None

# async def init_db():
#     global pool
#     pool = await asyncpg.create_pool(DATABASE_URL)

async def database_query_tool(input_data: QueryInput) -> Dict[str, Any]:
    """Execute database query with validation."""
    query = input_data.query
    params = input_data.params or []
    timeout = input_data.timeout

    start_time = datetime.now()
    logger.info("Executing database query")

    # Validate query type (prevent destructive operations if needed)
    forbidden_keywords = ["DROP", "TRUNCATE", "DELETE FROM"]
    upper_query = query.upper()

    for keyword in forbidden_keywords:
        if keyword in upper_query:
            raise ValueError(f"Query contains forbidden keyword: {keyword}")

    try:
        # TODO: Implement actual database query
        # async with pool.acquire() as conn:
        #     result = await conn.fetch(query, *params, timeout=timeout)

        # Placeholder implementation
        result = []

    duration = (datetime.now() - start_time).total_seconds()
    record_tool_call("database_query", duration, True)
    logger.info(f"Query completed in {duration:.3f}s")

        return {
            "rows": [dict(row) for row in result] if result else [],
            "row_count": len(result),
            "duration": 0.0
        }
    except Exception as e:
        raise ValueError(f"Database query failed: {str(e)}")

database_tool_schema = {
    "name": "database_query",
    "description": "Execute database queries with validation (configure connection first)"
}
